[iOS] FilterVC 리팩토링 하였습니다.

ehgud0670 opened the pull request:
### FilterVC 리팩토링 
> Issue #163 을 구현하였습니다.

* FilterViewController 뷰 구현 부분을 코드로 재구현하였습니다.
* Calendar VC,  Price VC가 Filter VC를 상속받는 구조로 리팩토링하였습니다. 
  * 따라서 Calendar VC 이면 타이틀로 "체크인 — 체크아웃"를 , Price VC이면 타이틀로 "가격" 을 보여주도록 했습니다.
* 가격 화면은 구현하지 않을 거라서 가격 버튼은 제거하였습니다.

> 구현 화면

<img width="250" alt="날짜" src="https://user-images.githubusercontent.com/38216027/85018180-b7d3fe80-b1a7-11ea-92a1-de9c30783e84.png">

<img width="250" alt="가격" src="https://user-images.githubusercontent.com/38216027/85018221-c7534780-b1a7-11ea-80f6-087ef7372ec2.png">

ehgud0670 committed:
Revert "Revert "[#160] refactor: RoomUseCase 의 세마포어 객체 삭제""

This reverts commit 9f64adde1d7458bbc1f64b127ef07af7814cf71e.

ehgud0670 committed:
[#163] refactor: 필터화면의 Header 뷰와 Footer 뷰를 코드로 구현

ehgud0670 committed:
이유: 코드로 구현할거라서 제거한다.

ehgud0670 committed:
[#163] refactor: 필터 스토리보드의 설정들 제거 및 stackView 설정을 코드로 옮김

ehgud0670 committed:
[#163] refactor: Instantiable 프로토콜 구현해 VC가 storyboard에 있든 없든 생성되도록 기능 수정
	이유: FilterViewController가 스토리보드로 불러올 필요가 없어져서 이렇게 재적용함

ehgud0670 committed:
[#163] refactor: FiterVC를 상속받는 VC를 만들어 상속을 이용해 뷰컨을 띄우도록 코드 수정

godrm commented on iOS/Airbnb/Airbnb/Controllers/FilterViewController.swift:
```diff
@@ -9,55 +9,76 @@
-    var filterType: FilterType?
+    deinit {
+        headerView.closeButton.removeTarget(self, action: #selector(close), for: .touchUpInside)
+    }
```
deinit 시점은 객체가 소멸되기 직전이라 FilterViewController 외부에 다른 객체를 참조하거나 연결된 것이 아니면 굳이 제거하지 않아도 모두 같이 사라집니다. headerView가 사라지지 않고 남아있어야 하는데 FilterViewController가 사라지기 때문에 관계를 끊어주는 건 아니겠죠?

ehgud0670 replied:
그럼 HeaderView.closeButton이 살아있는 경우에만, 그때 removeTarget을 하는거네요. 
결론적으로 만약 특정 UIControl 객체가 살아있는 경우라면 등록한 타겟을 deinit 될때 removeTarget을 해야 되는 거군요. 
사실 왜 removeTarget을 해야 하나 찾아본 적이 있는데 애플문서에서는 그 UIControl의 allTargets에 deinit될 타겟이 없어지니까 정리가 되는 차원에서 도움이 된다고 해서 그 정도만 알았지, 
removeTarget을 해야하는 특정 상황(해당 UIControl 객체가 계속 살아있는 경우)까지는 몰랐었는데 알려주셔서 감사합니다 ㅎㅎ 👍

ehgud0670 replied:
반영했습니다.

godrm commented on iOS/Airbnb/Airbnb/Controllers/FilterViewController.swift:
```diff
@@ -9,55 +9,76 @@
-        return viewController
-    }
-}
+extension FilterViewController: Instantiatable { }
```
구현 내용이 없는 데 프로토콜을 채택하는 게 어떤 의미가 있는걸까요? 
기존처럼 FilterViewController 에 구현해놓으면 상속받은 클래스는 모두 동일한 메소드를 사용할 수 있지 않을까요? Instantiatable 프로토콜을 다른 클래스에서 수평적으로 확장하는 데 많이 쓰나요? 

검색해보니 BNBTabBarController 에서도 채택을 하는군요. 그렇다면 이해는 됩니다. 
다만 구현부가 없이 채택만 하는 게 뭔가 다른 의도가 있는지 생각하게 되네요. 다른 방법은 없을까요?

ehgud0670 replied:
넵 그냥 채택만 하는 이유는 그저 수평적으로 사용하고 싶었기 때문입니다. 다형적인 용도는 아니었습니다. 
따라서 조언을 고려하여 그냥 VC를 생성시키는 객체인 `VCInstantiator`를 따로 생성해 특정 VC를 생성하도록 하였습니다. 
이 코드가 이전의 방법보다 더 좋은 이유는 따로 프로토콜을 채택할 필요가 없고 메소드도 수평적인 것이 확실해 보기에도 편하고 성능 최적화에도 좋은 것 같습니다.
객체의 역할도 분명히 보여서 더 좋구요. 이렇게 만들라고 하신거.. 맞죠?? ㅋㅋㅋ 😅

> 근데 객체 이름을 `ViewControllerInstantiator` 말고 `VCInstantiator`로 줄여쓴 이유는 너무 길어서 입니다. 이는 애플의 가이드라인을 무시하는 경우인데, 조금 길더라도 줄이지 말고 이름을 길게 적어야 되는건가요

godrm replied:
네. 수평적인 역할도 좋습니다. 다만 extension 자체가 어색하기는 합니다.

godrm commented on iOS/Airbnb/Airbnb/Controllers/RoomViewController.swift:
```diff
@@ -49,14 +49,42 @@ final class RoomViewController: UIViewController {
         }
     }
     
+    private func calendarViewController() -> CalendarViewController {
+        guard let calendarViewController = CalendarViewController.instantiate(
+            from: nil,
+            presentationStyle: .overCurrentContext,
+            transitionStyle: .crossDissolve
+            ) else { return CalendarViewController() }
+        
+        return calendarViewController
+    }
+    
+    private func priceViewController() -> PriceViewController {
+        guard let priceViewController = PriceViewController.instantiate(
+            from: nil,
+            presentationStyle: .overCurrentContext,
+            transitionStyle: .crossDissolve
+            ) else { return PriceViewController() }
+        
+        return priceViewController
+    }
```
여기 두 메소드는 제네릭 함수로 만들면 어떨까요?

ehgud0670 replied:
넵 감사합니다. 제네릭가 잠깐 머리에 스쳤는데 당연히 안된다고 생각했는데 해결사였네요. 
네 맞습니다. FilterViewController 자체를 리턴하면 자식이 부모에게 계속 가려져서 달력 VC 와 가격 VC가 를 표현을 못해서 답답하고 있던 차에 그냥 메소드를 달력 만드는거 하나, 가격 만드는 거 하나해서 두개 만들었던 거였습니다. ㅋ..
제네릭 진짜 좋네요 감사합니다.

godrm commented on iOS/Airbnb/Airbnb/SceneDelegate.swift:
```diff
@@ -30,7 +35,11 @@ final class SceneDelegate: UIResponder, UIWindowSceneDelegate {
+            ) else { return }
         
         DispatchQueue.main.asyncAfter(deadline: .now() + DispatchTimeInterval.nanoseconds(1)) {
             searchViewController.present(loginViewController, animated: true)
```
`.now() + DispatchTimeInterval.nanoseconds(1) ` 이면 그냥 async로 처리해도 되는거 아닌까요?

ehgud0670 replied:
넵, 그 생각을 못했네요. 알려주셔서 감사합니다. 👍
그리고 제가 async적으로 코드를 구현한 이유는 괜히 말씀드리면 
```
020-06-19 12:54:46.636558+0900 Airbnb[6525:1239692] Unbalanced calls to begin/end appearance transitions for <Airbnb.BNBTabBarController: 0x7fdf1901fa00>.
```
이런 메시지가 나타나서 그랬습니다.

감사합니다!!

ehgud0670 replied:
반영했습니다.

ehgud0670 committed:
[#163] refactor: async 코드를 좀 더 상황에 맞게 리팩토링함.

ehgud0670 committed:
[#163] refactor: removeTarget 하는 코드 삭제

	이유: closeButton은 어차피 VC가 deinit될때 같이 없어지므로 removeTarget을 안해줘도 된다. 따라서 해당 코드를 삭제한다.

ehgud0670 committed:
[#163] refactor: 1. VC를 생성하는 객체를 따로 만들어 제네릭을 이용해 특정 VC 생성하도록 함
		 2. RoomVC에서 FilterVC를 만드는 제네릭을 이용한 메소드를 만들어 VCInstantiator를 이용해 틀정 필터 뷰컨(달력, 가격)을 생성하도록 코드 수정함

ehgud0670 committed:
[#163] style: 줄인말은 가이드라인을 어기므로 약어를 다 풀어씀

godrm commented:
수고하셨습니다.
